### DISCUSSION TITLE
Run other recipe?

### DISCUSSION THREAD

**Discussion by fgblomqvist** [casey/just#818] [Q&A] [answered]
If I have a file like:
```
foo:
  echo foo

bar:
  echo bar
  foo
```

It will error because there is no such command as `foo`. Makes sense. But how can I call a recipe from another? I don't want to have to use `just foo` since working directory etc. isn't guaranteed.

**Comment by casey** [comment, marked as answer]
It isn't very pretty, but you can do:

```
foo:
  echo foo

bar:
  echo bar
  just --justfile {{justfile()}} foo
```

(`-f` can also be used instead of `--justfile`.)

This has come up a few times before, so ideally there would be a nice way to do this without such hackery.

**Reply by casey** [reply]
I opened #819 to track this, with some ideas for the syntax.

**Reply by fgblomqvist** [reply]
Hmm maybe I should have clarified a bit more. Another case is this:

```
foo:
  echo foo

bar:
  echo bar
  foo
  echo baz
```

As in, I want to be able to run another recipe at any time, however many times I want, and in any order.

**Reply by fgblomqvist** [reply]
Maybe for what I'm trying to do it's better to just split the code out in bash scripts and call those scripts from just

**Reply by casey** [reply]
It's a little trickier to run recipes in the middle of another recipe. Just doesn't do any parsing of the contents of recipes, it just treats it as a block of text. So it would be hard to add some kind of directive in the middle of a recipe.

So for example:

```
bar:

foo:
  echo a
  ??? bar
  echo b
```

I'm not sure what `???` could be that wouldn't conflict with a potentially valid recipe line.

One possibility is making just export some environment variables:

```
JUST_JUSTFILE # path to the current justfile
JUST_WORKING_DIRECTORY # path to the working directory
JUST_INVOCATION_DIRECTORY # path to invocation directory
```

And then adding a flag to just, like `--recurse`, that would read these variables to set up the environment, and then you could do:

```
bar:

foo:
  echo a
  just --recurse bar
  echo b
```

And you would be guaranteed to get the same justfiles and directories. This wouldn't preserve settings (like `just a=b foo`), so I suppose those would also have to be exported and read by `just --recurse`.

**Reply by casey** [reply]
That would miss command line flags as well, so maybe it's not a great approach.

**Reply by casey** [reply]
It's weird, but a prefix could work:

```
bar:

foo:
  echo a
: bar
  echo b
```

**Reply by fgblomqvist** [reply]
Yeah I was gonna ask how the @ symbol works? Because if I write `@echo "Hello"` just decides not to print the command, which is not a valid bash syntax. So yeah sounds like coming up with some other line prefix should work?

**Reply by casey** [reply]
`@` and `-` are sort of odd ducks. I added them because make supports them, but I'm not totally sure that it's a good idea to more.

I was thinking of a prefix that would go before the indentation, not after it.

**Reply by fgblomqvist** [reply]
I lean more and more towards just using `just` as the argument parser and then storing most/all code in bash scripts. That way, you don't lose out on just lacking file inclusion support, and you don't have to worry about this feature either (being able to run other recipes). It's not 100% as elegant as just doing everything, but without file-inclusion support there's not much to do anyway. But yeah, perhaps it would still make sense to add this feature 🤷

**Reply by bcheidemann-kmbal** [reply]
Has any progress been made on this? I just (pun not intended) started using just, and this was one of the first things I tried to do. `just --justfile {{justfile()}} foo` doesn't work reliably because the user invoking the recipe may not have just installed globally (i.e. `npm install -D just-install`), or have it installed under a different name (i.e. `npm install -g rust-just`). Maybe something like `{{justexecutable()}} --justfile {{justfile()}} foo` could work if there's not already a solution?

EDIT: Just found https://github.com/casey/just?tab=readme-ov-file#just-executable

**Reply by sociopart** [reply]
Since nothing was done until now, you can evolve this hackery by using variables:

```
call_recipe := just_executable() + " --justfile=" + justfile()

bar:
  @echo "test"

foo:
  {{call_recipe}} bar
```

But more proper solution would be fine. Mostly it might help for interacting with private recipes.

**Reply by rikhuijzer** [reply]
From the [cross-platform.just example](https://github.com/casey/just/blob/master/examples/cross-platform.just), it looks like there is a `&& syntax` that works:
```just
second:
  @echo "second"

first: && second
  @echo "first"
```

```sh
$ just second
second

$ just first
first
second
```

**Reply by homeworkprod** [reply]
The `&& syntax` only seems to work for a single recipe.

For me (just version 1.38.0), this works:
```
validate: && validate-this
    echo "Done."

validate-this:
    echo "Validating this."
```

But this does not:
```
validate: && validate-this && validate-that
    echo "Done."

validate-this:
    echo "Validating this."

validate-that:
    echo "Validating that."
```

**Reply by casey** [reply]
All recipes after `&&` will be subsequents, so you only need one:

```just
validate: && validate-this validate-that
    echo "Done."
```

**Reply by homeworkprod** [reply]
@casey thanks, that works for me!

**Reply by stevenh** [reply]
> Since nothing was done until now, you can evolve this hackery by using variables:
> 
> ```
> call_recipe := just_executable() + " --justfile=" + justfile()
> 
> bar:
>   @echo "test"
> 
> foo:
>   {{call_recipe}} bar
> ```
> 
> But more proper solution would be fine. Mostly it might help for interacting with private recipes.

Not this doesn't work in a loop on windows due to slashes.

```just
call_recipe := just_executable() + " --justfile=" + justfile()

single *val='default':
    @echo "val = {{val}}"

multiple *vals='first second':
    for val in {{vals}}; do \
        {{call_recipe}} single $val; \
    done
```

```log
just multiple
for val in first second; do C:\Users\XXX\AppData\Local\Microsoft\WinGet\Packages\Casey.Just_Microsoft.Winget.Source_8wekyb3d8bbwe\just.exe --justfile=C:\Users\XXX\code\github.com\rocketsciencegg\ai-playgound\go-ollama-perf\Justfile $val; done
sh: line 1: C:UsersXXXAppDataLocalMicrosoftWinGetPackagesCasey.Just_Microsoft.Winget.Source_8wekyb3d8bbwejust.exe: command not found
sh: line 1: C:UsersXXXAppDataLocalMicrosoftWinGetPackagesCasey.Just_Microsoft.Winget.Source_8wekyb3d8bbwejust.exe: command not found
error: Recipe `multiple` failed on line 44 with exit code 127
```

**Reply by enbugger** [reply]
none of it works inside modules

**Comment by GitToby** [comment]
just for further clarification, the && separates the "pre" and "post" commands. Remember the `@` symbol suppresses the printing of the command.

```
before1:
    @echo 1

before2 my_var="2":
    @echo {{my_var}}

after1:
    @echo a

after2 my_var="b":
    @echo {{my_var}}

cmd: before1 before2 && after1 (after2 "c")
    @echo "hello world"
```

prints 

```
1
2
hello world
a
c
```
